Add support for `write.data.path` (#1611)
apache/iceberg-python: mkdocs/docs/configuration.md

| `write.object-storage.enabled`           | Boolean                            | True    | Enables the [`ObjectStoreLocationProvider`](configuration.md#object-store-location-provider) that adds a hash component to file paths. Note: the default value of `True` differs from Iceberg's Java implementation |
| `write.object-storage.partitioned-paths` | Boolean                            | True    | Controls whether [partition values are included in file paths](configuration.md#partition-exclusion) when object storage is enabled                                                                                 |
| `write.py-location-provider.impl`        | String of form `module.ClassName`  | null    | Optional, [custom `LocationProvider`](configuration.md#loading-a-custom-location-provider) implementation                                                                                                           |

> | `write.data.path`                        | String pointing to location        | ∅        | Sets the location where to write the data. If not set, it will use the table location postfixed with `data/`.                                                                                                       |

Reviewer: Maybe something like:

```suggestion
| `write.data.path`                        | String referencing a location        | ∅        | Sets the location under which data is written. If not set, the table location (`table.metadata.location`) postfixed with `data/` is used.                                                                                                       |
```

Reviewer: Also, what about 'table location + /data' as the default instead of ∅ like in the Java docs?

Reviewer: Also bringing attention to https://github.com/apache/iceberg-python/pull/1537#discussion_r1921100230 - it would be great to make the location providers docs consistent with `write.data.path` now.

https://github.com/apache/iceberg-python/blob/4a055d786ae997921f8ac0c8393f6ae7d4fe41bb/mkdocs/docs/configuration.md?plain=1#L213

https://github.com/apache/iceberg-python/blob/4a055d786ae997921f8ac0c8393f6ae7d4fe41bb/mkdocs/docs/configuration.md?plain=1#L242

I think these are the only places that require changing - they'll be `write.data.path` (and maybe for clarity we should reiterate there that this defaults to `{location}/data` so examples are clearer). Could we do that in this PR maybe? (Happy to follow-up with this otherwise)

Author: Thanks for the input here, I've changed some wording en reworked some of the sections. Let me know what you think 👍

Reviewer: Looks great - thank you!

resulting in slowdowns. The `ObjectStoreLocationProvider` counteracts this by injecting deterministic hashes, in the form of binary directories,
into file paths, to distribute files across a larger number of object store prefixes.

> Paths still are also prefixed by `{location}/data/`, where `location` comes from the [table metadata](https://iceberg.apache.org/spec/#table-metadata-fields), in a similar manner to the [`SimpleLocationProvider`](configuration.md#simple-location-provider). This can be overridden by setting [`write.data.path` table configuration](#write-options).

Reviewer 2: ```suggestion
Paths are still prefixed by `{location}/data/`, where `location` comes from the [table metadata](https://iceberg.apache.org/spec/#table-metadata-fields), in a similar manner to the [`SimpleLocationProvider`](configuration.md#simple-location-provider). This can be overridden by setting [`write.data.path` table configuration](#write-options).
```
